nucliadb/nucliadb/purge/orphan_shards.py

```diff
@@ -0,0 +1,252 @@
+            extra={"nodes": list(unavailable_nodes)},
+        )
+
+    return orphan_shards
+
+
+async def _get_indexed_shards() -> dict[str, ShardLocation]:
+    indexed_shards: dict[str, ShardLocation] = {}
+    available_nodes = manager.get_index_nodes()
+    for node in available_nodes:
+        node_shards = await node.list_shards()
+        for shard_id in node_shards:
+            indexed_shards[shard_id] = ShardLocation(kbid=UNKNOWN_KB, node_id=node.id)
+    return indexed_shards
+
+
+async def _get_stored_shards(driver: Driver) -> dict[str, ShardLocation]:
+    stored_shards: dict[str, ShardLocation] = {}
+    shards_manager = KBShardManager()
+
+    async with driver.transaction(read_only=True) as txn:
+        async for kbid, _ in KnowledgeBox.get_kbs(txn, slug=""):
+            try:
+                kb_shards = await shards_manager.get_shards_by_kbid(kbid)
+            except ShardsNotFound:
+                logger.warning(
+                    "KB not found while looking for orphan shards", extra={"kbid": kbid}
+                )
+                continue
+            else:
+                for shard_object_pb in kb_shards:
+                    for shard_replica_pb in shard_object_pb.replicas:
+                        shard_replica_id = shard_replica_pb.shard.id
+                        node_id = shard_replica_pb.node
+                        stored_shards[shard_replica_id] = ShardLocation(
+                            kbid=kbid, node_id=node_id
+                        )
+    return stored_shards
+
+
+async def report_orphan_shards(
+    shards: dict[str, ShardLocation],
+    driver: Driver,
+):
+    async with driver.transaction(read_only=True) as txn:
+        for shard_id, location in shards.items():
+            kb_exists = await KnowledgeBox.exist_kb(txn, location.kbid)
+            if kb_exists:
+                msg = "Found orphan shard for existing KB"
+            else:
+                msg = "Found orphan shard for already removed KB"
+
+            logger.debug(
+                msg,
+                extra={
+                    "shard_id": shard_id,
+                    "kbid": location.kbid,
+                    "node_id": location.node_id,
+                },
+            )
+
+
+async def purge_orphan_shards(driver: Driver):
+    orphan_shards = await detect_orphan_shards(driver)
+
+    unavailable_nodes: dict[str, datetime] = {}
+    for shard_id, location in orphan_shards.items():
+        node = manager.get_index_node(location.node_id)
+        if node is None:
+            now = datetime.now()
+            last_try = unavailable_nodes.setdefault(location.node_id, now)
+            if (now - last_try).seconds < 10:
+                logger.warning(
+                    "Node not available while purging orphan shards, skipping",
+                    extra={"node_id": location.node_id},
+                )
+                unavailable_nodes[location.node_id] = now
+            continue
```

Comment:
the retry when the node is None logic seems a bit off, isn't it?
You could instead abstract into a `get_available_node` that handles the waiting and retry ?

Reply:
Simplified